src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1482 @@
+
+        let bytes_needed = HEADER_SIZE + data_bytes as usize + oob_bytes as usize;
+
+        // remove this message's bytes from read_buf, or just take read_buf
+        // if it contains exactly one message
+        let msg_buf = if self.read_buf.len() == bytes_needed {
+            mem::replace(&mut self.read_buf, Vec::with_capacity(READ_BUFFER_SIZE))
+        } else {
+            self.read_buf.drain(0..bytes_needed).collect::<Vec<u8>>()
+        };
+        dd!("[$ {:?}:{:?}] consumed {} bytes, left {}", self.iocp, self.handle, bytes_needed, self.read_buf.len());
+
+        // We have a full message available.
+        let data_end = HEADER_SIZE + data_bytes as usize;
+        let mut buf_data = msg_buf[HEADER_SIZE..data_end].to_vec();
+
+        let mut channels: Vec<OsOpaqueIpcChannel> = vec![];
+        let mut shmems: Vec<OsIpcSharedMemory> = vec![];
+
+        if oob_bytes > 0 {
+            let buf_oob = &msg_buf[data_end..];
+            let msg = bincode::serde::deserialize::<OsIpcOutOfBandMessage>(&buf_oob).unwrap();
+
+            assert!(msg.channel_is_sender.len() == msg.channel_handles.len());
+            assert!(msg.shmem_sizes.len() == msg.shmem_handles.len());
+
+            dd2!("[$ {:?}:{:?}] msg with total {} bytes, {} channels, {} shmems, big data handle 0x{:x}",
+                 self.iocp, self.handle, data_bytes,
+                 msg.channel_handles.len(), msg.shmem_handles.len(),
+                 msg.big_data_receiver_handle);
+
+            for (handle, is_sender) in msg.channel_handles.iter().zip(msg.channel_is_sender.iter()) {
+                channels.push(OsOpaqueIpcChannel::new(*is_sender, *handle as HANDLE));
+            }
+
+            for (handle, size) in msg.shmem_handles.iter().zip(msg.shmem_sizes.iter()) {
+                shmems.push(OsIpcSharedMemory::from_handle(*handle as HANDLE, *size as usize).unwrap());
+            }
+
+            if msg.big_data_receiver_handle != 0 {
+                dd2!("[$ {:?}:{:?}] read msg with big data, rec handle {:?}", self.iocp, self.handle, msg.big_data_receiver_handle as HANDLE);
+                let receiver = OsIpcReceiver::from_handle(msg.big_data_receiver_handle as HANDLE);
+                let big_msg = try!(receiver.recv().map_err(|err| panic!("Failed to receive subchannel big data: {:?}", err)));
+                buf_data = big_msg.0;
+            }
+        }
+
+        dd!("[$ {:?}:{:?}] get_message success -> {} bytes, {} channels, {} shmems",
+            self.iocp, self.handle, buf_data.len(), channels.len(), shmems.len());
+        Ok((buf_data, channels, shmems))
+    }
+}
+
+#[derive(Debug)]
+pub struct OsIpcReceiver {
+    // The handle to the receiver end of a pipe
+    handle: WinHandle,
+
+    // The handle to the sender end of a pipe, if it hasn't been taken
+    // yet; None if it has.
+    sender: Option<OsIpcSender>,
+
+    // The IOCP that this is a part of.  A bare HANDLE, since we don't
+    // own it.
+    iocp: HANDLE,
```

Comment:
(BTW, sorry if I'm too severe with these things: if this was one of the back-ends I can test myself, I could just submit cleanup PRs myself after it's merged -- but since I can't do that here, I'd prefer it to be as elegant as possible from the start... Of course you are free to reject my notions of "elegant" in cases where it mostly boils down to stylistic preferences :-) )